File: CodeGenDotNet 159059-1372 (a)/CSharp/Component Shells/WinSupport/DateTimeKADUC.cs
// ====================================================================
//  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
// ====================================================================
//   I'm distributing this code so you'll be able to use it to see code
//   generation in action and I hope it will be useful and you'll enjoy 
//   using it. This code is provided "AS IS" without warranty, either 
//   expressed or implied, including implied warranties of merchantability 
//   and/or fitness for a particular purpose. 
//  ====================================================================
//  Summary: Very simple user control to support null handling of dates

using System;

namespace KADGen.WinSupport
{
	public class DateTimeUC : System.Windows.Forms.UserControl
	{

		private System.Drawing.Color mOriginalForeColor;
		private System.Windows.Forms.DateTimePickerFormat mOriginalFormat;
		private string mOriginalCustomFormat;
		private bool mReadOnly;

		#region Windows Form Designer generated code

		public DateTimeUC() : base()
		{
			//This call is required by the Windows Form Designer.
			InitializeComponent();

			//Add any initialization after the InitializeComponent() call

		}

		//UserControl overrides dispose to clean up the component list.
		protected  override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if( ! (components == null) )
				{
					components.Dispose();
				}
			}
			base.Dispose(disposing);
		}

		//Required by the Windows Form Designer
		private System.ComponentModel.IContainer components;

		//NOTE: The following procedure is required by the Windows Form Designer
		//It can be modified using the Windows Form Designer.  
		//Do not modify it using the code editor.
		internal System.Windows.Forms.CheckBox chkEmpty;
		internal System.Windows.Forms.DateTimePicker dtsValue ;
		internal  System.Windows.Forms.TextBox txtReadOnlyValue;
		[System.Diagnostics.DebuggerStepThrough()] private void InitializeComponent()
		{
			this.chkEmpty = new System.Windows.Forms.CheckBox();
			this.txtReadOnlyValue = new System.Windows.Forms.TextBox();
			this.dtsValue = new System.Windows.Forms.DateTimePicker();
			this.SuspendLayout();
			//
			//chkEmpty
			//
			this.chkEmpty.CheckAlign = System.Drawing.ContentAlignment.MiddleRight;
			this.chkEmpty.Location = new System.Drawing.Point(0 , 0);
			this.chkEmpty.Name = "chkEmpty";
			this.chkEmpty.Size = new System.Drawing.Size(56 , 20);
			this.chkEmpty.TabIndex = 0;
			this.chkEmpty.Text = "Empty";
			this.chkEmpty.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
			this.chkEmpty.CheckedChanged += new System.EventHandler(this.chkEmpty_CheckedChanged);
			//
			//txtReadOnlyValue
			//
			this.txtReadOnlyValue.Enabled = false;
			this.txtReadOnlyValue.Location = new System.Drawing.Point(0 , 0);
			this.txtReadOnlyValue.Name = "txtReadOnlyValue";
			this.txtReadOnlyValue.ReadOnly = true;
			this.txtReadOnlyValue.Size = new System.Drawing.Size(264 , 20);
			this.txtReadOnlyValue.TabIndex = 2;
			this.txtReadOnlyValue.Text = "";
			this.txtReadOnlyValue.Visible = false;
			//
			//dtsValue
			//
			this.dtsValue.Format = System.Windows.Forms.DateTimePickerFormat.Short;
			this.dtsValue.Location = new System.Drawing.Point(88 , 0);
			this.dtsValue.Name = "dtsValue";
			this.dtsValue.Size = new System.Drawing.Size(176 , 20);
			this.dtsValue.TabIndex = 3;
			this.dtsValue.CloseUp += new System.EventHandler(this.dtsValue_CloseUp);
			this.dtsValue.TextChanged += new System.EventHandler(this.dtsValue_TextChanged);
			//
			//DateTimeUC
			//
			this.Controls.Add(this.dtsValue);
			this.Controls.Add(this.chkEmpty);
			this.Controls.Add(this.txtReadOnlyValue);
			this.Name = "DateTimeUC";
			this.Size = new System.Drawing.Size(264 , 20);
			this.ResumeLayout(false);

		}

		#endregion

		#region Public Properties and Methods
		public override string Text 
		{
			get
			{
				if( chkEmpty.Checked )
				{
					return "";
				}
				else
				{
					return dtsValue.Text;
				}
			}
			set
			{
				if( (value == "") | (value == "1/1/1900") )
				{
					chkEmpty.Checked = true;
					dtsValue.Width = dtsValue.Height;
					txtReadOnlyValue.Text = "";
				}
				else
				{
					chkEmpty.Checked = false;
					dtsValue.Width = dtsValue.Height;
					dtsValue.Text = value;
					txtReadOnlyValue.Text = value;
				}
			}
		}

		public string EmptyText
		{
			get
			{
				return chkEmpty.Text;
			}
			set
			{
				chkEmpty.Text = value;
			}
		}

		public System.Windows.Forms.Appearance CheckAppearance
		{
			get
			{
				return chkEmpty.Appearance;
			}
			set
			{
				chkEmpty.Appearance = value;
			}
		}

		public System.Windows.Forms.DateTimePickerFormat Format 
		{
			get
			{
				return dtsValue.Format;
			}
			set
			{
				dtsValue.Format = value;
			}
		}

		public string CustomFormat
		{ 
			get
			{
				return dtsValue.CustomFormat;
			}
			set
			{
				dtsValue.CustomFormat = value;
			}
		}

		public System.DateTime MinDate 
		{
			get
			{
				return dtsValue.MinDate;
			}
			set
			{
				dtsValue.MinDate = value;
			}
		}

		public System.DateTime MaxDate 
		{
			get
			{
				return dtsValue.MaxDate;
			}
			set
			{
				dtsValue.MaxDate = value;
			}
		}

		public bool @ReadOnly
		{
			get
			{
				return mReadOnly;
			}
			set
			{
				mReadOnly = value;
				chkEmpty.Visible = ! value;
				dtsValue.Visible = ! value;
				txtReadOnlyValue.Visible = value;
			}
		}

		#endregion

		#region Event Handlers
		private void chkEmpty_CheckedChanged( object sender, System.EventArgs e )
		{
			if( this.ActiveControl != dtsValue )
			{
				this.PerformLayout();
			}
		}

		private void dtsValue_TextChanged(object sender, System.EventArgs e )
		{
			chkEmpty.Checked = false;
			OnTextChanged(e);
		}

		private void dtsValue_CloseUp(object sender, System.EventArgs e )
		{
			this.PerformLayout();
		}

		protected override void OnLayout(System.Windows.Forms.LayoutEventArgs levent )
		{
			if( mReadOnly )
			{
				dtsValue.Left = 0;
				dtsValue.Width = this.Width;
			}
			else
			{
				dtsValue.Left = chkEmpty.Right;
				dtsValue.Width = this.Width - chkEmpty.Right;
			}
			if( ! this.DesignMode )
			{
				if( this.mOriginalFormat != 0 )
				{
					if( chkEmpty.Checked )
					{
						dtsValue.ForeColor = System.Drawing.Color.AntiqueWhite;
						dtsValue.Format = System.Windows.Forms.DateTimePickerFormat.Custom;
						dtsValue.CustomFormat = "////";
					}
					else
					{
						dtsValue.Format = this.mOriginalFormat;
						dtsValue.CustomFormat = this.mOriginalCustomFormat;
						dtsValue.ForeColor = mOriginalForeColor;
					}
				}
			}
		}

		protected override void OnLoad( System.EventArgs e )
		{
			this.mOriginalForeColor = this.ForeColor;
			this.mOriginalCustomFormat = this.CustomFormat;
			this.mOriginalFormat = this.Format;
		}

		#endregion
	}
}

File: CodeGenDotNet 159059-1372 (a)/VBCode/Component Shells/WinSupport/DateTimeKADUC.vb
' ====================================================================
'  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
' ====================================================================
'   I'm distributing this code so you'll be able to use it to see code
'   generation in action and I hope it will be useful and you'll enjoy 
'   using it. This code is provided "AS IS" without warranty, either 
'   expressed or implied, including implied warranties of merchantability 
'   and/or fitness for a particular purpose. 
'  ====================================================================
'  Summary: Very simple user control to support null handling of dates

Option Strict On
Option Explicit On 

Imports System

Public Class DateTimeUC
   Inherits System.Windows.Forms.UserControl

   Private mOriginalForeColor As System.Drawing.Color
   Private mOriginalFormat As Windows.Forms.DateTimePickerFormat
   Private mOriginalCustomFormat As String
   Private mReadOnly As Boolean

#Region " Windows Form Designer generated code "

   Public Sub New()
      MyBase.New()

      'This call is required by the Windows Form Designer.
      InitializeComponent()

      'Add any initialization after the InitializeComponent() call

   End Sub

   'UserControl overrides dispose to clean up the component list.
   Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
      If disposing Then
         If Not (components Is Nothing) Then
            components.Dispose()
         End If
      End If
      MyBase.Dispose(disposing)
   End Sub

   'Required by the Windows Form Designer
   Private components As System.ComponentModel.IContainer

   'NOTE: The following procedure is required by the Windows Form Designer
   'It can be modified using the Windows Form Designer.  
   'Do not modify it using the code editor.
   Friend WithEvents chkEmpty As System.Windows.Forms.CheckBox
   Friend WithEvents dtsValue As System.Windows.Forms.DateTimePicker
   Friend WithEvents txtReadOnlyValue As System.Windows.Forms.TextBox
   <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
      Me.chkEmpty = New System.Windows.Forms.CheckBox
      Me.txtReadOnlyValue = New System.Windows.Forms.TextBox
      Me.dtsValue = New System.Windows.Forms.DateTimePicker
      Me.SuspendLayout()
      '
      'chkEmpty
      '
      Me.chkEmpty.CheckAlign = System.Drawing.ContentAlignment.MiddleRight
      Me.chkEmpty.Location = New System.Drawing.Point(0, 0)
      Me.chkEmpty.Name = "chkEmpty"
      Me.chkEmpty.Size = New System.Drawing.Size(56, 20)
      Me.chkEmpty.TabIndex = 0
      Me.chkEmpty.Text = "Empty"
      Me.chkEmpty.TextAlign = System.Drawing.ContentAlignment.MiddleCenter
      '
      'txtReadOnlyValue
      '
      Me.txtReadOnlyValue.Enabled = False
      Me.txtReadOnlyValue.Location = New System.Drawing.Point(0, 0)
      Me.txtReadOnlyValue.Name = "txtReadOnlyValue"
      Me.txtReadOnlyValue.ReadOnly = True
      Me.txtReadOnlyValue.Size = New System.Drawing.Size(264, 20)
      Me.txtReadOnlyValue.TabIndex = 2
      Me.txtReadOnlyValue.Text = ""
      Me.txtReadOnlyValue.Visible = False
      '
      'dtsValue
      '
      Me.dtsValue.Format = System.Windows.Forms.DateTimePickerFormat.Short
      Me.dtsValue.Location = New System.Drawing.Point(88, 0)
      Me.dtsValue.Name = "dtsValue"
      Me.dtsValue.Size = New System.Drawing.Size(176, 20)
      Me.dtsValue.TabIndex = 3
      '
      'DateTimeUC
      '
      Me.Controls.Add(Me.dtsValue)
      Me.Controls.Add(Me.chkEmpty)
      Me.Controls.Add(Me.txtReadOnlyValue)
      Me.Name = "DateTimeUC"
      Me.Size = New System.Drawing.Size(264, 20)
      Me.ResumeLayout(False)

   End Sub

#End Region

#Region "Public Properties and Methods"
   Public Overrides Property Text() As String
      Get
         If chkEmpty.Checked Then
            Return ""
         Else
            Return dtsValue.Text
         End If
      End Get
      Set(ByVal Value As String)
         If Value = "" OrElse Value = "1/1/1900" Then
            chkEmpty.Checked = True
            dtsValue.Width = dtsValue.Height
            txtReadOnlyValue.Text = ""
         Else
            chkEmpty.Checked = False
            dtsValue.Width = dtsValue.Height
            dtsValue.Text = Value
            txtReadOnlyValue.Text = Value
         End If
      End Set
   End Property

   Public Property EmptyText() As String
      Get
         Return chkEmpty.Text
      End Get
      Set(ByVal Value As String)
         chkEmpty.Text = Value
      End Set
   End Property

   Public Property CheckAppearance() As System.Windows.Forms.Appearance
      Get
         Return chkEmpty.Appearance
      End Get
      Set(ByVal Value As System.Windows.Forms.Appearance)
         chkEmpty.Appearance = Value
      End Set
   End Property

   Public Property Format() As Windows.Forms.DateTimePickerFormat
      Get
         Return dtsValue.Format
      End Get
      Set(ByVal Value As Windows.Forms.DateTimePickerFormat)
         dtsValue.Format = Value
      End Set
   End Property

   Public Property CustomFormat() As String
      Get
         Return dtsValue.CustomFormat
      End Get
      Set(ByVal Value As String)
         dtsValue.CustomFormat = Value
      End Set
   End Property

   Public Property MinDate() As System.DateTime
      Get
         Return dtsValue.MinDate
      End Get
      Set(ByVal Value As System.DateTime)
         dtsValue.MinDate = Value
      End Set
   End Property

   Public Property MaxDate() As System.DateTime
      Get
         Return dtsValue.MaxDate
      End Get
      Set(ByVal Value As System.DateTime)
         dtsValue.MaxDate = Value
      End Set
   End Property

   Public Property [ReadOnly]() As Boolean
      Get
         Return mReadOnly
      End Get
      Set(ByVal Value As Boolean)
         mReadOnly = Value
         chkEmpty.Visible = Not Value
         dtsValue.Visible = Not Value
         txtReadOnlyValue.Visible = Value
      End Set
   End Property

#End Region

#Region "Event Handlers"
   Private Sub chkEmpty_CheckedChanged( _
               ByVal sender As Object, _
               ByVal e As System.EventArgs) _
               Handles chkEmpty.CheckedChanged
      'If chkEmpty.Checked Then
      '   dtsValue.Checked = True
      '   'dtsValue.Text = ""
      '   ' set dts empty
      'Else
      '   dtsValue.Checked = True
      '   ' enable dts
      'End If
      If Not Me.ActiveControl Is dtsValue Then
         Me.PerformLayout()
      End If
   End Sub

   Private Sub dtsValue_TextChanged( _
               ByVal sender As System.Object, _
               ByVal e As System.EventArgs) _
               Handles dtsValue.TextChanged
      chkEmpty.Checked = False
      onTextChanged(e)
   End Sub

   'Private Sub dtsValue_Validated(ByVal sender As Object, ByVal e As System.EventArgs) Handles dtsValue.Validated
   '   onvalidated(e)
   'End Sub

   Private Sub dtsValue_CloseUp( _
               ByVal sender As Object, _
               ByVal e As System.EventArgs) _
               Handles dtsValue.CloseUp
      Me.PerformLayout()
   End Sub

   'Private Sub dtsValue_ValueChanged( _
   '            ByVal sender As System.Object, _
   '            ByVal e As System.EventArgs) '_
   '            Handles dtsValue.ValueChanged
   '   chkEmpty.Checked = (dtsValue.Text <> "")
   'End Sub

   Protected Overrides Sub OnLayout( _
               ByVal levent As System.Windows.Forms.LayoutEventArgs)
      If mReadOnly Then
         dtsValue.Left = 0
         dtsValue.Width = Me.Width
      Else
         dtsValue.Left = chkEmpty.Right
         dtsValue.Width = Me.Width - chkEmpty.Right
      End If
      If Not Me.DesignMode Then
         If Not Me.mOriginalFormat = 0 Then
            If chkEmpty.Checked Then
               dtsValue.ForeColor = System.Drawing.Color.AntiqueWhite
               dtsValue.Format = Windows.Forms.DateTimePickerFormat.Custom
               dtsValue.CustomFormat = "''"
               ' dtsValue.Text()
            Else
               dtsValue.Format = Me.mOriginalFormat
               dtsValue.CustomFormat = Me.mOriginalCustomFormat
               dtsValue.ForeColor = mOriginalForeColor
            End If
         End If
      End If
   End Sub

   Protected Overrides Sub OnLoad(ByVal e As System.EventArgs)
      Me.mOriginalForeColor = Me.ForeColor
      Me.mOriginalCustomFormat = Me.CustomFormat
      Me.mOriginalFormat = Me.Format
   End Sub

#End Region


End Class